google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryCountTest.java

```diff
@@ -0,0 +1,336 @@
+    CollectionReference collection = createCollectionWithDocuments(5);
+    AggregateQuerySnapshot snapshot = collection.count().get().get();
+    assertThat(snapshot.getCount()).isEqualTo(5);
+  }
+
+  @Test
+  public void countShouldReturnNumMatchingDocuments() throws Exception {
+    CollectionReference collection = createEmptyCollection();
+    createDocumentsWithKeyValuePair(collection, 3, "key", 42);
+    createDocumentsWithKeyValuePair(collection, 5, "key", 24);
+    AggregateQuerySnapshot snapshot = collection.whereEqualTo("key", 42).count().get().get();
+    assertThat(snapshot.getCount()).isEqualTo(3);
+  }
+
+  @Test
+  public void countShouldRespectLimit() throws Exception {
+    CollectionReference collection = createCollectionWithDocuments(5);
+    AggregateQuerySnapshot snapshot = collection.limit(2).count().get().get();
+    assertThat(snapshot.getCount()).isEqualTo(2);
+  }
+
+  @Test
+  public void countShouldRespectOffset() throws Exception {
+    CollectionReference collection = createCollectionWithDocuments(5);
+    AggregateQuerySnapshot snapshot = collection.offset(2).count().get().get();
+    assertThat(snapshot.getCount()).isEqualTo(3);
+  }
+
+  @Test
+  public void countShouldRespectOrderBy() throws Exception {
+    CollectionReference collection = createEmptyCollection();
+    createDocumentsWithKeyValuePair(collection, 3, "key1", 42);
+    createDocumentsWithKeyValuePair(collection, 5, "key2", 24);
+    createDocumentsWithKeyValuePair(collection, 4, "key1", 99);
+    AggregateQuerySnapshot snapshot = collection.orderBy("key1").count().get().get();
+    // Note: A subtle side effect of order-by is that it filters out documents that do not have the
+    // order-by field.
+    assertThat(snapshot.getCount()).isEqualTo(7);
+  }
+
+  @Test
+  public void countShouldRespectStartAtAndEndAt() throws Exception {
+    CollectionReference collection = createCollectionWithDocuments(10);
+    List<QueryDocumentSnapshot> documentSnapshots = collection.get().get().getDocuments();
+    AggregateQuerySnapshot snapshot =
+        collection
+            .startAt(documentSnapshots.get(2))
+            .endAt(documentSnapshots.get(7))
+            .count()
+            .get()
+            .get();
+    assertThat(snapshot.getCount()).isEqualTo(6);
+  }
+
+  @Test
+  public void countShouldReturnNumberOfDocumentsForCollectionGroups() throws Exception {
+    CollectionGroup collectionGroup = createCollectionGroupWithDocuments(13);
+    AggregateQuerySnapshot snapshot = collectionGroup.count().get().get();
+    assertThat(snapshot.getCount()).isEqualTo(13);
+  }
+
+  @Test
+  public void inFlightCountQueriesShouldCompleteSuccessfullyWhenFirestoreIsClosed()
+      throws Exception {
+    CollectionReference collection = createCollectionWithDocuments(20);
+    ApiFuture<AggregateQuerySnapshot> task = collection.count().get();
+    collection.getFirestore().close();
```

Comment:
Interesting, will this be flaky?

Reply:
So if for whatever reason the backend does not respond, `close()` will hang forever?

Reply:
Well, I would like to see it for my curiosity now. Not a blocker for this PR though.